### PR TITLE
Close #23804: Use version instead of date for telemetry/metrics expiration

### LINKED ISSUE #23804: Use version instead of date for telemetry/metrics expiration
We currently use a fixed date for telemetry expiration which has numerous downsides:

    Pings can expire anytime within a release cycle. We had this happen for search telemetry which dropped off a few days after a dot release. This prompted the change to set business critical metrics to never expire.

    It's technically impossible to achieve reproducible builds using expiration by date. There's always a point in time at which a release/tag can no longer be checked out and shipped. Telemetry tests would fail, or telemetry would simply not be sent in case there's no test coverage. And urgent dot release for a security or stability issue would then require going through completely unrelated telemetry renewal before it can be shipped.

    Telemetry renewals are difficult to carry out as the expiration dates aren't aligned with release dates.

Switching to expiration by version addresses the concerns above. Firefox Desktop already switched to version based expiry and this is supported by Glean. 

See also: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1724760
https://firefox-source-docs.mozilla.org/toolkit/components/glean/user/new_definitions_file.html#how-does-expiry-work

### PR DESCRIPTION
**Pull request by rocketsroger**
This change is created by using a script with the following conversion table
_CONVERSION = {
  '    expires: "2022-05': '    expires: 101',
  '    expires: "2022-06': '    expires: 102',
  '    expires: "2022-07': '    expires: 103',
  '    expires: "2022-08': '    expires: 104',
  '    expires: "2022-09': '    expires: 105',
  '    expires: "2022-10': '    expires: 106',
  '    expires: "2022-11': '    expires: 107',
  '    expires: "2022-12': '    expires: 108',
  '    expires: "2023-01': '    expires: 109',
  '    expires: "2023-02': '    expires: 111',
  '    expires: "2023-03': '    expires: 112',
  '    expires: "2023-04': '    expires: 113',
  '    expires: "2023-05': '    expires: 114',
  '    expires: "2023-06': '    expires: 115',
  '    expires: "2023-07': '    expires: 116',
  '    expires: "2023-08': '    expires: 117',
  '    expires: "2023-09': '    expires: 118',
  '    expires: "2023-10': '    expires: 119',
  '    expires: "2023-11': '    expires: 121',
  '    expires: "2023-12': '    expires: 122',
  '    expires: "2024-02': '    expires: 123',
  '    expires: "2024-03': '    expires: 124',
  '    expires: "2024-04': '    expires: 125',
  '    expires: "2024-05': '    expires: 126',
  '    expires: "2024-06': '    expires: 127',
  '    expires: "2024-07': '    expires: 128',
}

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/metrics.yaml
Comment (csadilek) [inline app/metrics.yaml:160]: Glean docs say "X" (where X is the major version...), and use expires: "88" as an example. Is it fine to skip the quotes?

**Review by csadilek — COMMENTED** [review]
Thanks for moving this forward! Let's reach out to the Glean team for a review here to double-check.

File: app/metrics.yaml
Comment (rocketsroger) [reply app/metrics.yaml:160]: Yes, we don't need to put the quotes around the version number.  I got the example from the Glean team here: https://github.com/mozilla/glean/blob/main/samples/android/app/metrics.yaml

**Comment by rocketsroger** [comment]
> Thanks for moving this forward! Let's reach out to the Glean team for a review here to double-check.

Thanks! I'll add the Glean team to the review.

File: app/metrics.yaml
Comment (csadilek) [reply app/metrics.yaml:160]: ok, perfect!

Comment (badboy) [reply app/metrics.yaml:160]: Woops, guess we didn't update the docs then

**Comment by Dexterp37** [comment]
Removing me from the reviewers list, @badboy is taking care of it. Please make sure he gets to review _before_ merging :-)

File: app/build.gradle
Comment (Dexterp37) [inline app/build.gradle:841]: This will expire all your metrics: you need to provide the major version of Fenix here. See the [Glean docs](https://mozilla.github.io/glean/book/language-bindings/android/android-build-configuration-options.html#gleanexpirebyversion)

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: app/build.gradle
Comment (rocketsroger) [reply app/build.gradle:841]: I see, this version tells Glean what is the major version.  So with 1 set then no metrics will expire, correct?

Comment (Dexterp37) [reply app/build.gradle:841]: @rocketsroger correct, but I strongly discourage you from landing it as it is: you should really provide the correct Fenix version in here, within this PR. If you want non-expiring metrics, you can make non-expiring metrics via the metrics.yaml (by requesting the proper data-review).

Comment (rocketsroger) [reply app/build.gradle:841]: @Dexterp37 yes, I won't land this until I update the patch.  Just want to make sure I understand this correctly.  The version number we pass in here is the version number Glean will compare against the metrics `expires:` field correct?  So if we pass in `ext.gleanExpireByVersion = 99` then anything below 100 will expire but metrics that `expires: 100` should still work.  Correct?

File: app/build.gradle
Comment (rocketsroger) [inline app/build.gradle:841]: Not sure if this is the best way to get the major version.  It is the simplest way though.

File: app/build.gradle
Comment (Dexterp37) [reply app/build.gradle]: `gleanExpireByVersion = 99` means that Fenix is version 99.x.x so all the metrics bound to expire in versions <= 99 will be expired.

File: app/build.gradle
Comment (csadilek) [reply app/build.gradle:841]: @rocketsroger I think we can add a method specifically for `Fenix` that will work with both the Nightly and Release versioning schemes. Happy to help with that.

**Review by Dexterp37 — APPROVED** [review]
This looks good to me from a data collectio POV. Please make sure a Fenix peer reviews the way you are fetching the major version before merging :-)

**Review by csadilek — APPROVED** [review]
Tested with Debug/Nightly and Release builds. All looks good to me.

One TODO we have left is to update the wiki and document our new renewal process, as discussed. On merge day, when we upgrade to the next Nightly version, we want to trigger renewal for metrics expiring in that version + 1.

**Comment by Dexterp37** [comment]
> One TODO we have left is to update the wiki and document our new renewal process, as discussed. On merge day, when we upgrade to the next Nightly version, we want to trigger renewal for metrics expiring in that version + 1.

@csadilek that's interesting, can I hear more about why do you have this process in place? Will you automatically renew all metrics like that or just the subset you'd still need? Why bump just by one version? (did you consider making them never expiring?)

**Comment by csadilek** [comment]
@Dexterp37 We're not bumping by one version, that was ambiguous/incorrect wording on my part after Roger and I had already talked in more detail. We will start **reviewing** (not _renewing_ automatically) all metrics that expire in that version + 1. So when we switch to a new Nightly version we will also look at the metrics that expire in this version and the next to make sure nothing falls through. The rest stays the same. We will still check with product what's still needed and start renewal where needed.

**Pull request merged**
